### ISSUE TITLE
upper case all device serial numbers when they come in

### ISSUE THREAD

**Issue by sungo** [joyent/conch#513]
Humans are entering serial numbers in mixed case, which creates multiple devices when one is intended or causes the UI to throw an unexpected error.  To make this better for everyone, let's upper-case all device serials when we using them in the API.

**Comment by karenetheridge** [comment]
Would that include rewriting the device's id in the incoming report before we store it? (otherwise, deduping comparisons will fail on the different device id.)

**Comment by karenetheridge** [comment]
It would probably be better if device ids were upper-cased closer to the source -- e.g. by the ui itself when the operator typed it in - so the id that is saved on disk and used for reports is already the "correct" one.  (and ditto for stripping whitespace.)

**Comment by karenetheridge** [comment]
paging @perigrin for that last item - can we employ some simple js wizardry to either reject the device id on input if it has spaces/lowercase characters, or do the munging as it goes by?

**Comment by sungo** [comment]
I'd prefer we do this in the API so that every client doesn't have to implement the same logic.

**Comment by karenetheridge** [comment]
I will assume the desire is to normalize device ids on *all* endpoints, not just as device reports are processed and the device record is created.

**Comment by karenetheridge** [comment]
Before we can roll out any code changes, we need to resolve the existing bad entries:

`select concat('--',id,'--') from device where id ~ '(^\s|\s$)' or id != upper(id);`

-> one entry with trailing whitespace
-> 19 entries with lower-cased ids

And we can't just do this:

`update device set id = regexp_replace(upper(id), '(^\s+|\s+$)', '') where id ~ '(^\s+|\s+$)' or id != upper(id);`

...because some of these ids are already duplicated by their upper-cased counterparts -- therefore the related database entries need to be merged/removed first. This probably requires surgery by hand.

**Comment by karenetheridge** [comment]
I have a migration script prepared that deals with the issues I found with a db backup made a few months ago, but this will need to be reviewed closer to launch date against the current production database in case there are other bad device ids added in the meantime.

**Comment by karenetheridge** [comment]
I would go further and limit device ids (serials) to just `/^[A-Z0-9_-]+$/`.  There are just a few ids with a `+` or `)` present, which look like typos.

There are a dozen or so with `-`, e.g. 'AMS-PRD5', which we could possibly also rename as well if we want to stick to `/^[A-Z0-9_]+$/`.

**Comment by sungo** [comment]
@daleghent what are your thoughts here about potential device IDs / serial numbers and their possible character range?

**Comment by sungo** [comment]
paging @jemershaw and @bdelano as well

**Comment by karenetheridge** [comment]
This will be done in v3 as part of making device ids be a uuid (#474)

**Comment by jemershaw** [comment]
I would prefer if the serials contained dashes or had the ability to toggle that setting. When I do development work with virtual machines the serial ends up being a uuid from the vm which contains dashes. I'm assuming this change will still allow that to work?

**Comment by sungo** [comment]
The more I think about it, the more I think this shouldn't be an API concern. The API should not be altering data like this. Now, we could put requirements on device IDs in general via our json schemas. That'd prevent API users from submitting bad data. However, my original ask was for the API to alter the device serial numbers so I'm going to close this out. We can open another ticket for json schema limitations should we decide that's a good idea.
